lib/Dialect/Ibis/Transforms/IbisContainersToHW.cpp

```diff
@@ -0,0 +1,306 @@
+//===- IbisContainersToHW.cpp ---------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/Ibis/IbisDialect.h"
+#include "circt/Dialect/Ibis/IbisOps.h"
+#include "circt/Dialect/Ibis/IbisPasses.h"
+#include "circt/Dialect/Ibis/IbisTypes.h"
+
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace circt;
+using namespace ibis;
+
+static llvm::StringRef getScopeRefModuleName(Type type) {
+  return type.cast<ScopeRefType>().getScopeRef().getAttr().strref();
+}
+
+namespace {
+
+// Analysis result for generating the port interface of a container + a bit of
+// port op caching.
+struct ContainerPortInfo {
+  std::unique_ptr<hw::ModulePortInfo> hwPorts;
+
+  // A mapping between the port name and the port op within the container.
+  llvm::DenseMap<StringAttr, InputPortOp> opInputs;
+
+  // A mapping between the port name and the port op within the container.
+  llvm::DenseMap<StringAttr, OutputPortOp> opOutputs;
+
+  ContainerPortInfo() = default;
+  ContainerPortInfo(ContainerOp container) {
```

Comment:
nit: This seems like a bit much for a constructor. I prefer keeping constructors to simple initialization (constant time) and anything else in a method... but I don't know how standard that is and I'm not aware of any comments in the LLVM style guide about it.